### PR TITLE
Added real head tracking using a gamepad device

### PR DESCRIPTION
**Pull request by hoene**
Gamepads are the easiest way of getting rotation data into the browser...

### CODE REVIEW

**Comment by hoch** [comment]
@hoene Sorry for the late response. Thanks for your contribution Christian! Could you change few things before we start reviewing the PR?

1. Please use Google JS style guide.
2. Using a game pad device is a great idea and it deserves a separate file. Please create a new file for your demo!
3. What if user does not have a game pad device? The demo should notify that the game pad is required. We don't have to be fancy and I guess alert() is fine for that matter.

**Comment by hoene** [comment]
I signed the CLA.

File: examples/gamepad.html [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+-->
+<!DOCTYPE html>
+<html>
+<head>
+  <meta charset="utf-8">
+  <title>A Simple Demo: Omnitone with Headtracking</title>
```
Comment (hoch) [inline examples/gamepad.html]: Headtracking -> Headtracking via gamepad

File: examples/gamepad.html [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+      }, function (onInitializationError) {
+        console.error(onInitializationError);
+      });
+
+      var textRotation = document.getElementById('rotation');
+
+      function runAnimation() {
+
+        // look for a Gamepad and check, whether is has changed it axes values.
+        var gamepads = navigator.getGamepads ? navigator.getGamepads() : (navigator.webkitGetGamepads ? navigator.webkitGetGamepads() : []);
```
Comment (hoch) [inline examples/gamepad.html]: Let's wrap the code at 80 col.

```js
var gamepads = navigator.getGamepads 
    ? navigator.getGamePads()
    : (navigator.webkitGetGamepads ? navigator.webkitGetGamepads : []);
```

What happens if `gamepads = []`?

Also does this have to be done every render call? Can we just cache the gamepad object upfront and use it?

File: examples/gamepad.html [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+        var changed = false;
+        for (var i = 0; i < gamepads.length; ++i) {
+          var pad = gamepads[i];
+          if (pad != null) {
+            if (roll != pad.axes[2] || pitch != pad.axes[1] || yaw != pad.axes[0]) {
+              changed = true;
+              yaw = pad.axes[0];
+              pitch = pad.axes[1];
+              roll = pad.axes[2];
+              textRotation.innerHTML = "yaw " + Math.round(yaw * 180)
```
Comment (hoch) [inline examples/gamepad.html]: Google style prefers single quote to double quote.

File: examples/gamepad.html [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+                + " pitch " + Math.round(pitch * 180) 
+                + " roll " + Math.round(roll * 180);
+            }
+            break;
+          }
+        }
+
+        if (changed) {
+          // from http://mathworld.wolfram.com/EulerAngles.html
+          // down is back
```
Comment (hoch) [inline examples/gamepad.html]: Where does this comment belong? The line 134?

File: examples/gamepad.html [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+          
+          var phi = yaw * Math.PI; 
+          // 1. the first rotation is by an angle phi about the z-axis using D,
```
Comment (hoch) [inline examples/gamepad.html]: Please adjust comment location. It's slightly confusing.

File: examples/gamepad.html [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+          
+          var theta = pitch * Math.PI; 
+          // the second rotation is by an angle theta in [0,pi] about the former x-axis (now x^') using C, and
```
Comment (hoch) [inline examples/gamepad.html]: This comment needs to be wrapped at 80.

File: examples/gamepad.html [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+    }
+
+    select {
+      width: 120px;
+      font-size: 24px;
+    }
+  </style>
+</head>
+<body>
+  <h2>Select the sound source to play.</h2>
```
Comment (hoch) [inline examples/gamepad.html]: We might want to add a header with `<h1>` to say this demo is about the gamepad.

File: examples/gamepad.html [hunk 8]
```diff
@@ -0,0 +1,162 @@
+<!--
+  Copyright 2016 Google Inc. All Rights Reserved.
```
Comment (hoch) [inline examples/gamepad.html]: 2016 -> 2017

File: examples/gamepad.html [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+          audioElement.play();
+      };
+
+      var rotator;
+      
+      // Initialize the decoder and activate the button when ready.
+      decoder.initialize().then(function () {
+        buttonPlayback.disabled = false;
+        selectSource.disabled = false;
+        rotator = Omnitone.createFOARotator(audioContext);
```
Comment (hoch) [inline examples/gamepad.html]: This is not necessary. The decode already has a rotator inside.

File: examples/gamepad.html [hunk 10]
```diff
@@ -0,0 +1,162 @@
...
+      };
+
+      selectSource.onchange = function (event) {
+        audioElement.src = audioSources[event.target.value];
+        audioElement.load();
+        if (buttonPlayback.textContent === 'Pause')
+          audioElement.play();
+      };
+
+      var rotator;
```
Comment (hoch) [inline examples/gamepad.html]: We don't need this. See the comment below.

File: examples/gamepad.html [hunk 11]
```diff
@@ -0,0 +1,162 @@
...
+          var phi = yaw * Math.PI; 
+          // 1. the first rotation is by an angle phi about the z-axis using D,
+          
+          var theta = pitch * Math.PI; 
+          // the second rotation is by an angle theta in [0,pi] about the former x-axis (now x^') using C, and
+          
+          var psi = roll * Math.PI;   
+          //3. the third rotation is by an angle psi about the former z-axis (now z^') using B.
+
+          rotator.setRotationMatrix([
```
Comment (hoch) [inline examples/gamepad.html]: Replace `rotator` with `decoder`

File: examples/gamepad.html [hunk 12]
```diff
@@ -0,0 +1,162 @@
...
+            Math.cos(psi) * Math.sin(phi) + Math.cos(theta) * Math.cos(phi) * Math.sin(psi),
+            Math.sin(psi) * Math.sin(theta),
+            -Math.sin(psi) * Math.cos(phi) - Math.cos(theta) * Math.sin(phi) * Math.cos(psi),
+            -Math.sin(psi) * Math.sin(phi) + Math.cos(theta) * Math.cos(phi) * Math.cos(psi),
+            Math.cos(psi) * Math.sin(theta),
+            Math.sin(theta) * Math.sin(phi),
+            -Math.sin(theta) * Math.cos(phi),
+            Math.cos(theta)
+          ]);
+          decoder.setRotationMatrix(rotator.getRotationMatrix());
```
Comment (hoch) [inline examples/gamepad.html]: ... and remove this line.

File: examples/gamepad.html [hunk 13]
```diff
@@ -92,68 +124,95 @@
...
-            Math.cos(theta)
-          ]);
-          decoder.setRotationMatrix(rotator.getRotationMatrix());
+        if (gamepad) {
+          divRotator.style.display = 'block';
+          selectRotation(0);
+          gamepad = false;
         }
-        window.requestAnimationFrame(runAnimation);
+        setTimeout(runAnimation, 500);
```
Comment (hoch) [inline examples/gamepad.html]: Hmm. Why is this needed?

File: examples/gamepad.html [hunk 14]
```diff
@@ -92,68 +124,95 @@
...
+      // check checks for gamepads or query the existing gamepad for changes
       function runAnimation() {
 
         // look for a Gamepad and check, whether is has changed it axes values.
-        var gamepads = navigator.getGamepads ? navigator.getGamepads() : (navigator.webkitGetGamepads ? navigator.webkitGetGamepads() : []);
-        var changed = false;
+        var gamepads = navigator.getGamepads ? navigator.getGamepads() :
+          (navigator.webkitGetGamepads ? navigator.webkitGetGamepads()
+            : []);
         for (var i = 0; i < gamepads.length; ++i) {
```
Comment (hoch) [inline examples/gamepad.html]: This looks like you're handling multiple gamepad in a single update. So if there are multiple pads connected and actively moving, what's going to happen?

File: examples/gamepad.html [hunk 15]
```diff
@@ -92,68 +124,95 @@
...
           if (pad != null) {
-            if (roll != pad.axes[2] || pitch != pad.axes[1] || yaw != pad.axes[0]) {
-              changed = true;
-              yaw = pad.axes[0];
-              pitch = pad.axes[1];
-              roll = pad.axes[2];
-              textRotation.innerHTML = "yaw " + Math.round(yaw * 180) 
-                + " pitch " + Math.round(pitch * 180) 
-                + " roll " + Math.round(roll * 180);
+            if (!gamepad) {
```
Comment (hoch) [inline examples/gamepad.html]: What's the purpose of this check?

File: examples/gamepad.html [hunk 16]
```diff
@@ -92,68 +124,95 @@
...
       }, function (onInitializationError) {
         console.error(onInitializationError);
       });
 
-      var textRotation = document.getElementById('rotation');
+      var textRotation = document.getElementById('eRotation');
+      var divRotator = document.getElementById('eRotator');
 
+      // called if HTML selector is changed
+      selectRotation = function (value) {
```
Comment (hoch) [inline examples/gamepad.html]: This variable is not declared previously.

Comment (hoene) [inline examples/gamepad.html]: yes, I'll fix it.

File: examples/gamepad.html (hunk 13 above)
Comment (hoene) [inline examples/gamepad.html]: It is just to save CPU time. If no gamepad is present, then you do not have to update that frequent.

File: examples/gamepad.html (hunk 14 above)
Comment (hoene) [inline examples/gamepad.html]: I just look for the first gamepad that provides suitable data and then stop.

File: examples/gamepad.html (hunk 15 above)
Comment (hoene) [inline examples/gamepad.html]: I changed the name to gamepadFound to make it more clearly. If you remove the gamepad, then the selector is displayed again.

File: examples/gamepad.html (hunk 12 above)
Comment (hoene) [inline examples/gamepad.html]: done

File: examples/gamepad.html (hunk 11 above)
Comment (hoene) [inline examples/gamepad.html]: done
